File: kernel/rust/src/fs/tarfs.rs
//! Read-only TAR filesystem (supports .tar and .tar.gz).

use alloc::boxed::Box;
use alloc::collections::BTreeMap;
use alloc::string::String;
use alloc::vec::Vec;

use crate::vfs::traits::{
    CallerId, DirEntry, FileHandle, FileSystem, FsError, KPath, Metadata, NodeType, OpenFlags,
    Result, SeekFrom,
};

/// TAR entry types (POSIX ustar typeflag byte)
const TAR_TYPE_HARDLINK: u8 = 0x31; // '1' - hard link to another archived entry
const TAR_TYPE_SYMLINK: u8 = 0x32; // '2' - symbolic link (target in linkname field)
const TAR_TYPE_DIR: u8 = 0x35; // '5' - directory

/// An entry in the tar archive
#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct TarEntry {
    pub path: String,
    pub data_offset: usize, // Offset into tar_data where file content starts
    pub size: usize,
    pub entry_type: NodeType,
    pub mode: u32,
    pub mtime: i64,
    /// Hard-link count (POSIX `st_nlink`): the number of archive names that
    /// resolve to this inode's bytes. `1` for an unlinked file; computed for
    /// hard-link groups after indexing (entries sharing a `data_offset`).
    pub nlink: u32,
    /// Symbolic-link target text (verbatim, for `Symlink` entries); empty
    /// otherwise. Hard-link targets are resolved to bytes at index time.
    pub target: String,
}

/// Read-only TAR filesystem
#[allow(dead_code)]
pub struct TarFs {
    /// Raw tar data (could be compressed or uncompressed)
    data: Vec<u8>,
    /// Index of entries by path
    entries: BTreeMap<String, TarEntry>,
    /// Is the data gzip compressed?
    compressed: bool,
}

impl TarFs {
    /// Create a new TarFs from tar data
    /// If the data starts with gzip magic, it will be treated as compressed
    pub fn new(data: Vec<u8>) -> Result<Self> {
        let compressed = data.len() >= 2 && data[0] == 0x1f && data[1] == 0x8b;

        let mut fs = Self {
            data,
            entries: BTreeMap::new(),
            compressed,
        };

        fs.build_index()?;

        // Ensure root directory exists
        if !fs.entries.contains_key("/") {
            fs.entries.insert(
                String::from("/"),
                TarEntry {
                    path: String::from("/"),
                    data_offset: 0,
                    size: 0,
                    entry_type: NodeType::Dir,
                    mode: 0o755,
                    mtime: 0,
                    nlink: 1,
                    target: String::new(),
                },
            );
        }

        Ok(fs)
    }

    /// Parse a TAR header and extract info
    fn parse_header(&self, offset: usize) -> Option<(TarEntry, usize)> {
        if offset + 512 > self.data.len() {
            return None;
        }

        let header = &self.data[offset..offset + 512];

        // Check for end-of-archive (two zero blocks)
        if header.iter().all(|&b| b == 0) {
            return None;
        }

        // Read name (bytes 0-99) plus the optional ustar prefix (345-499).
        let name = self.read_path(header);

        // Read size (bytes 124-135, octal)
        let size_str = self.read_string(header, 124, 12);
        let size = u64::from_str_radix(&size_str.trim(), 8).unwrap_or(0) as usize;

        // Read type flag (byte 156)
        let typeflag = header.get(156).copied().unwrap_or(0x30);
        let entry_type = match typeflag {
            TAR_TYPE_DIR => NodeType::Dir,
            TAR_TYPE_SYMLINK => NodeType::Symlink,
            _ => NodeType::File,
        };

        // Link target (bytes 157-256). A symlink stores its target verbatim; a
        // hard link names another archived entry (normalized to a path here and
        // resolved to that entry's bytes in `build_index`).
        let linkname = self.read_string(header, 157, 100);
        let target = match typeflag {
            TAR_TYPE_SYMLINK => linkname,
            TAR_TYPE_HARDLINK => tar_path(&linkname),
            _ => String::new(),
        };

        // Read mode (bytes 100-107, octal)
        let mode_str = self.read_string(header, 100, 8);
        let mode = u32::from_str_radix(&mode_str.trim(), 8).unwrap_or(0o644);

        // Read mtime (bytes 136-147, octal)
        let mtime_str = self.read_string(header, 136, 12);
        let mtime = i64::from_str_radix(&mtime_str.trim(), 8).unwrap_or(0);

        // Calculate data offset (after header, rounded up to 512)
        let data_offset = offset + 512;
        let next_offset = data_offset + ((size + 511) / 512) * 512;

        let path = tar_path(&name);

        let entry = TarEntry {
            path: path.clone(),
            data_offset,
            size,
            entry_type,
            mode,
            mtime,
            nlink: 1,
            target,
        };

        Some((entry, next_offset))
    }

    /// Read a null-terminated string from header
    fn read_string(&self, header: &[u8], start: usize, max_len: usize) -> String {
        let end = (start..start + max_len)
            .find(|&i| header.get(i).copied() == Some(0))
            .unwrap_or(start + max_len);

        String::from(String::from_utf8_lossy(&header[start..end]))
    }

    /// Read a ustar path from `name` (0-99) and the optional `prefix` (345-499).
    /// The `prefix` field only exists in STRICT POSIX ustar headers (magic
    /// `"ustar\0"`, which `utils` emits); GNU headers (the xtask base image)
    /// reuse bytes 345+ for atime/ctime, so reading a prefix there would corrupt
    /// the path. Gate on the magic: GNU/non-ustar headers use the name field only
    /// (their long paths, if any, ride a separate GNU long-name record).
    fn read_path(&self, header: &[u8]) -> String {
        let name = self.read_string(header, 0, 100);
        let posix_ustar = header.len() >= 263 && &header[257..263] == b"ustar\0";
        if !posix_ustar {
            return name;
        }
        let prefix = self.read_string(header, 345, 155);
        if prefix.is_empty() {
            name
        } else if name.is_empty() {
            prefix
        } else {
            alloc::format!("{prefix}/{name}")
        }
    }

    /// Build index of all entries
    fn build_index(&mut self) -> Result<()> {
        let mut offset = 0usize;

        while offset < self.data.len() {
            if let Some((mut entry, next_offset)) = self.parse_header(offset) {
                // Resolve a hard link ('1') to the bytes of its (already-indexed)
                // target so both names read the same content.
                if entry.entry_type == NodeType::File && !entry.target.is_empty() {
                    if let Some(target) = self.entries.get(&entry.target) {
                        entry.data_offset = target.data_offset;
                        entry.size = target.size;
                    }
                    entry.target = String::new();
                }
                self.entries.insert(entry.path.clone(), entry);
                offset = next_offset;
            } else {
                break;
            }
        }

        // Hard-link counts: file entries that share a resolved `data_offset` are
        // the same inode (a hard-link group from `utils::append_hardlink`), so
        // report `st_nlink` = the group size. Only regular files participate.
        let mut links: BTreeMap<usize, u32> = BTreeMap::new();
        for entry in self.entries.values() {
            if entry.entry_type == NodeType::File {
                *links.entry(entry.data_offset).or_insert(0) += 1;
            }
        }
        for entry in self.entries.values_mut() {
            if entry.entry_type == NodeType::File {
                if let Some(&n) = links.get(&entry.data_offset) {
                    entry.nlink = n;
                }
            }
        }

        Ok(())
    }

    /// Get entry by path
    fn get_entry(&self, path: &str) -> Option<&TarEntry> {
        self.entries.get(path)
    }

    /// Normalize path
    fn normalize_path(&self, path: &KPath) -> String {
        let s = path.as_str();
        if s.is_empty() || s == "." {
            String::from("/")
        } else {
            String::from(s)
        }
    }
}

/// File handle for reading from tar
#[allow(dead_code)]
pub struct TarFileHandle {
    path: String,
    offset: u64,
    entry: TarEntry,
    data: *const Vec<u8>,
}

impl FileHandle for TarFileHandle {
    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
        let start = self.offset as usize;
        let data_offset = self.entry.data_offset;
        let size = self.entry.size;

        if start >= size {
            return Ok(0);
        }

        let end = (start + buf.len()).min(size);
        let to_read = end - start;

        unsafe {
            let data = &*self.data;
            buf[..to_read].copy_from_slice(&data[data_offset + start..data_offset + end]);
        }

        self.offset += to_read as u64;
        Ok(to_read)
    }

    fn write(&mut self, _buf: &[u8]) -> Result<usize> {
        Err(FsError::PermissionDenied) // Read-only filesystem
    }

    fn seek(&mut self, pos: SeekFrom) -> Result<u64> {
        let size = self.entry.size as i64;
        let new_offset = match pos {
            SeekFrom::Start(n) => n as i64,
            SeekFrom::Current(n) => self.offset as i64 + n,
            SeekFrom::End(n) => size + n,
        };

        if new_offset < 0 {
            return Err(FsError::InvalidPath);
        }

        self.offset = new_offset as u64;
        Ok(self.offset)
    }

    fn stat(&self) -> Result<Metadata> {
        // A handle is only ever created over a regular file (open rejects
        // directories and symlinks).
        Ok(entry_meta(&self.entry))
    }
}

/// Build `Metadata` from a tar entry, surfacing the archived permission bits and
/// mtime (tar stores mtime in **seconds**; we use ms). The archive is read-only,
/// so `atime = ctime = mtime`.
fn entry_meta(entry: &TarEntry) -> Metadata {
    let ms = entry.mtime.saturating_mul(1000);
    let base = match entry.entry_type {
        NodeType::Dir => Metadata::dir(),
        // Report the real hard-link count so a committed hard-link group surfaces
        // as `st_nlink > 1` (e.g. `stat` "Links: 2") after a stacked boot.
        NodeType::File => Metadata::file_with_nlink(entry.size as u64, entry.nlink),
        NodeType::Symlink => Metadata::symlink(entry.target.len() as u64),
    };
    base.with_mode((entry.mode & 0o7777) as u16)
        .with_times(ms, ms, ms)
}

/// Normalize a tar entry name to an absolute VFS path: strip a leading `./`,
/// ensure a leading `/`, and drop any trailing `/`.
fn tar_path(name: &str) -> String {
    if name.starts_with("./") {
        let stripped = name[1..].trim_end_matches('/');
        if stripped.is_empty() {
            String::from("/")
        } else {
            String::from(stripped)
        }
    } else if name.starts_with('/') {
        String::from(name.trim_end_matches('/'))
    } else {
        alloc::format!("/{}", name.trim_end_matches('/'))
    }
}

impl FileSystem for TarFs {
    fn open(
        &mut self,
        _caller: CallerId,
        path: &KPath,
        flags: OpenFlags,
    ) -> Result<Box<dyn FileHandle>> {
        let path_str = self.normalize_path(path);

        // Check if file exists
        let entry = self
            .get_entry(&path_str)
            .cloned()
            .ok_or(FsError::NotFound)?;

        // Check if trying to write
        if flags.write || flags.create || flags.truncate {
            return Err(FsError::PermissionDenied); // Read-only
        }

        // Directories and symlinks are not opened as byte streams (a symlink
        // would have been resolved by the namespace before reaching here).
        match entry.entry_type {
            NodeType::Dir => return Err(FsError::IsDir),
            NodeType::Symlink => return Err(FsError::InvalidPath),
            NodeType::File => {}
        }

        let handle = TarFileHandle {
            path: path_str,
            offset: 0,
            entry,
            data: &self.data as *const Vec<u8>,
        };

        Ok(Box::new(handle))
    }

    fn stat(&self, path: &KPath) -> Result<Metadata> {
        let path_str = self.normalize_path(path);
        let entry = self.get_entry(&path_str).ok_or(FsError::NotFound)?;
        Ok(entry_meta(entry))
    }

    fn readdir(&self, _caller: CallerId, path: &KPath) -> Result<Vec<DirEntry>> {
        let path_str = self.normalize_path(path);

        // Find entries that are direct children of this path
        let prefix = if path_str.ends_with('/') {
            path_str.clone()
        } else {
            alloc::format!("{}/", path_str)
        };

        let mut result = Vec::new();

        for (entry_path, entry) in &self.entries {
            if entry_path == &path_str {
                continue; // Skip self
            }

            if entry_path.starts_with(&prefix) {
                // Get the relative name
                let relative = &entry_path[prefix.len()..];

                // Only include direct children (no slashes in relative path)
                if !relative.contains('/') {
                    result.push(DirEntry {
                        name: String::from(relative),
                        node_type: entry.entry_type,
                    });
                }
            }
        }

        Ok(result)
    }

    fn mkdir(&mut self, _caller: CallerId, _path: &KPath) -> Result<()> {
        Err(FsError::PermissionDenied) // Read-only
    }

    fn unlink(&mut self, _caller: CallerId, _path: &KPath) -> Result<()> {
        Err(FsError::PermissionDenied) // Read-only
    }

    fn rename(&mut self, _caller: CallerId, _from: &KPath, _to: &KPath) -> Result<()> {
        Err(FsError::PermissionDenied) // Read-only
    }

    fn readlink(&self, path: &KPath) -> Result<String> {
        let path_str = self.normalize_path(path);
        let entry = self.get_entry(&path_str).ok_or(FsError::NotFound)?;
        if entry.entry_type == NodeType::Symlink {
            Ok(entry.target.clone())
        } else {
            Err(FsError::InvalidPath)
        }
    }
}

File: memcontainers/kernel/zig/src/fs/tarfs.zig
//! src/fs/tarfs.zig — read-only POSIX-ustar base-image view (§2.5).
//!
//! Owns: ustar parsing, hardlink/symlink treatment, read-only file/dir handles. The base
//! images are uncompressed .tar (matching the oracle, whose `compressed` flag never drives
//! inline decompression). Oracle: kernel/rust/src/fs/tarfs.rs.
//! Not here: COW writes (cowfs.zig); mount policy (vfs.zig).

const std = @import("std");
const vfs = @import("../vfs.zig");
const FsError = vfs.FsError;
const NodeType = vfs.NodeType;
const Metadata = vfs.Metadata;
const OpenFlags = vfs.OpenFlags;
const SeekFrom = vfs.SeekFrom;
const FileHandle = vfs.FileHandle;
const FileSystem = vfs.FileSystem;
const DirEntry = vfs.DirEntry;

const MAX_PATH = 4096;
const TAR_TYPE_HARDLINK: u8 = '1';
const TAR_TYPE_SYMLINK: u8 = '2';
const TAR_TYPE_DIR: u8 = '5';

const TarEntry = struct {
    data_offset: usize,
    size: usize,
    entry_type: NodeType,
    mode: u32,
    mtime: i64,
    nlink: u32,
    target: []const u8, // owned; symlink target text (empty for non-symlinks)
};

pub const TarFs = struct {
    gpa: std.mem.Allocator,
    data: []u8, // owned tar bytes
    entries: std.StringHashMapUnmanaged(TarEntry) = .{}, // owns path keys

    /// Build a TarFs over owned `data`. Returns null on a malformed archive.
    pub fn create(gpa: std.mem.Allocator, data: []u8) ?*TarFs {
        const self = gpa.create(TarFs) catch @panic("OOM");
        self.* = .{ .gpa = gpa, .data = data };
        self.buildIndex();
        if (!self.entries.contains("/")) {
            self.entries.put(gpa, gpa.dupe(u8, "/") catch @panic("OOM"), .{
                .data_offset = 0,
                .size = 0,
                .entry_type = .dir,
                .mode = 0o755,
                .mtime = 0,
                .nlink = 1,
                .target = "",
            }) catch @panic("OOM");
        }
        return self;
    }

    pub fn fileSystem(self: *TarFs) FileSystem {
        return .{ .ptr = self, .vtable = &fs_vtable };
    }

    fn readOctal(field: []const u8) u64 {
        const t = std.mem.trim(u8, field, " \x00");
        return std.fmt.parseInt(u64, t, 8) catch 0;
    }

    fn headerString(header: []const u8, start: usize, max_len: usize) []const u8 {
        var end = start;
        while (end < start + max_len and end < header.len and header[end] != 0) : (end += 1) {}
        return header[start..end];
    }

    /// ustar path: name (0-99) + optional strict-ustar prefix (345-499). GNU headers reuse
    /// bytes 345+ for times, so gate the prefix on the strict `ustar\0` magic (257-262).
    fn readPath(self: *TarFs, gpa: std.mem.Allocator, header: []const u8) []const u8 {
        _ = self;
        const name = headerString(header, 0, 100);
        const posix_ustar = header.len >= 263 and std.mem.eql(u8, header[257..263], "ustar\x00");
        if (!posix_ustar) return gpa.dupe(u8, name) catch @panic("OOM");
        const prefix = headerString(header, 345, 155);
        if (prefix.len == 0) return gpa.dupe(u8, name) catch @panic("OOM");
        if (name.len == 0) return gpa.dupe(u8, prefix) catch @panic("OOM");
        return std.fmt.allocPrint(gpa, "{s}/{s}", .{ prefix, name }) catch @panic("OOM");
    }

    fn buildIndex(self: *TarFs) void {
        var scratch = std.heap.ArenaAllocator.init(self.gpa);
        defer scratch.deinit();
        const sa = scratch.allocator();

        var offset: usize = 0;
        while (offset + 512 <= self.data.len) {
            const header = self.data[offset .. offset + 512];
            var all_zero = true;
            for (header) |b| {
                if (b != 0) {
                    all_zero = false;
                    break;
                }
            }
            if (all_zero) break;

            const raw_name = self.readPath(sa, header);
            const size: usize = @intCast(readOctal(header[124..136]));
            const typeflag = header[156];
            const entry_type: NodeType = switch (typeflag) {
                TAR_TYPE_DIR => .dir,
                TAR_TYPE_SYMLINK => .symlink,
                else => .file,
            };
            const linkname = headerString(header, 157, 100);
            const mode: u32 = @intCast(readOctal(header[100..108]));
            const mtime: i64 = @intCast(readOctal(header[136..148]));
            const data_offset = offset + 512;
            const next_offset = data_offset + ((size + 511) / 512) * 512;
            const path = tarPath(self.gpa, raw_name);

            var target: []const u8 = "";
            if (typeflag == TAR_TYPE_SYMLINK) {
                target = self.gpa.dupe(u8, linkname) catch @panic("OOM");
            }

            var entry = TarEntry{
                .data_offset = data_offset,
                .size = size,
                .entry_type = entry_type,
                .mode = if (mode == 0) 0o644 else mode,
                .mtime = mtime,
                .nlink = 1,
                .target = target,
            };
            // Resolve a hard link ('1') to its (already-indexed) target's bytes.
            if (entry_type == .file and typeflag == TAR_TYPE_HARDLINK) {
                const hl_target = tarPath(sa, linkname);
                if (self.entries.get(hl_target)) |t| {
                    entry.data_offset = t.data_offset;
                    entry.size = t.size;
                }
            }
            self.putEntry(path, entry);
            offset = next_offset;
        }

        // Hard-link counts: file entries sharing a resolved data_offset are one inode.
        var links = std.AutoHashMapUnmanaged(usize, u32){};
        defer links.deinit(self.gpa);
        var it = self.entries.valueIterator();
        while (it.next()) |e| {
            if (e.entry_type == .file) {
                const gop = links.getOrPut(self.gpa, e.data_offset) catch @panic("OOM");
                if (!gop.found_existing) gop.value_ptr.* = 0;
                gop.value_ptr.* += 1;
            }
        }
        var it2 = self.entries.valueIterator();
        while (it2.next()) |e| {
            if (e.entry_type == .file) {
                if (links.get(e.data_offset)) |n| e.nlink = n;
            }
        }
    }

    fn putEntry(self: *TarFs, path: []const u8, entry: TarEntry) void {
        const gop = self.entries.getOrPut(self.gpa, path) catch @panic("OOM");
        if (gop.found_existing) {
            self.gpa.free(path);
        }
        gop.value_ptr.* = entry;
    }

    fn getEntry(self: *TarFs, path_in: []const u8) ?*TarEntry {
        var nb: [MAX_PATH]u8 = undefined;
        const path = normPath(path_in, &nb) orelse return null;
        return self.entries.getPtr(path);
    }

    fn entryMeta(entry: *const TarEntry) Metadata {
        const ms = entry.mtime *| 1000;
        const base = switch (entry.entry_type) {
            .dir => Metadata.dir(),
            .file => Metadata.fileWithNlink(entry.size, entry.nlink),
            .symlink => Metadata.symlink(entry.target.len),
        };
        return base.withMode(@intCast(entry.mode & 0o7777)).withTimes(ms, ms, ms);
    }

    // ── FileSystem impl ───────────────────────────────────────────────────────────────────
    fn open(self: *TarFs, path: []const u8, flags: OpenFlags) FsError!FileHandle {
        const entry = self.getEntry(path) orelse return FsError.NotFound;
        if (flags.write or flags.create or flags.truncate) return FsError.PermissionDenied;
        switch (entry.entry_type) {
            .dir => return FsError.IsDir,
            .symlink => return FsError.InvalidPath,
            .file => {},
        }
        const h = self.gpa.create(TarFileHandle) catch @panic("OOM");
        h.* = .{ .fs = self, .offset = 0, .data_offset = entry.data_offset, .size = entry.size, .nlink = entry.nlink, .mode = entry.mode, .mtime = entry.mtime };
        return .{ .ptr = h, .vtable = &TarFileHandle.handle_vtable };
    }

    fn stat(self: *TarFs, path: []const u8) FsError!Metadata {
        const entry = self.getEntry(path) orelse return FsError.NotFound;
        return entryMeta(entry);
    }

    fn readdir(self: *TarFs, path_in: []const u8, arena: std.mem.Allocator, out: *std.ArrayList(DirEntry)) FsError!void {
        var nb: [MAX_PATH]u8 = undefined;
        const path = normPath(path_in, &nb) orelse return FsError.InvalidPath;
        var prefix_buf: [MAX_PATH + 1]u8 = undefined;
        const prefix = if (std.mem.endsWith(u8, path, "/")) path else blk: {
            @memcpy(prefix_buf[0..path.len], path);
            prefix_buf[path.len] = '/';
            break :blk prefix_buf[0 .. path.len + 1];
        };
        var it = self.entries.iterator();
        while (it.next()) |kv| {
            const ep = kv.key_ptr.*;
            if (std.mem.eql(u8, ep, path)) continue;
            if (std.mem.startsWith(u8, ep, prefix)) {
                const rel = ep[prefix.len..];
                if (std.mem.indexOfScalar(u8, rel, '/') == null) {
                    out.append(arena, .{ .name = arena.dupe(u8, rel) catch @panic("OOM"), .node_type = kv.value_ptr.entry_type }) catch @panic("OOM");
                }
            }
        }
    }

    fn readlink(self: *TarFs, path: []const u8, out: *std.ArrayList(u8)) FsError!void {
        const entry = self.getEntry(path) orelse return FsError.NotFound;
        if (entry.entry_type != .symlink) return FsError.InvalidPath;
        out.appendSlice(self.gpa, entry.target) catch @panic("OOM");
    }

    const fs_vtable = FileSystem.VTable{
        .open = fsOpen,
        .stat = fsStat,
        .readdir = fsReaddir,
        .mkdir = fsReadOnly1,
        .unlink = fsReadOnly1,
        .rename = vfs.fsRenameUnsupported,
        .symlink = vfs.fsSymlinkUnsupported,
        .link = vfs.fsLinkUnsupported,
        .readlink = fsReadlink,
        .setMode = vfs.fsSetModeUnsupported,
        .setTimes = vfs.fsSetTimesUnsupported,
    };
    fn fsOpen(p: *anyopaque, _: vfs.CallerId, path: []const u8, flags: OpenFlags) FsError!FileHandle {
        return self_(p).open(path, flags);
    }
    fn fsStat(p: *anyopaque, path: []const u8) FsError!Metadata {
        return self_(p).stat(path);
    }
    fn fsReaddir(p: *anyopaque, _: vfs.CallerId, path: []const u8, arena: std.mem.Allocator, out: *std.ArrayList(DirEntry)) FsError!void {
        return self_(p).readdir(path, arena, out);
    }
    fn fsReadOnly1(_: *anyopaque, _: vfs.CallerId, _: []const u8) FsError!void {
        return FsError.PermissionDenied;
    }
    fn fsReadlink(p: *anyopaque, path: []const u8, out: *std.ArrayList(u8)) FsError!void {
        return self_(p).readlink(path, out);
    }
    fn self_(p: *anyopaque) *TarFs {
        return @ptrCast(@alignCast(p));
    }
};

const TarFileHandle = struct {
    fs: *TarFs,
    offset: u64,
    data_offset: usize,
    size: usize,
    nlink: u32,
    mode: u32,
    mtime: i64,

    fn read(self: *TarFileHandle, buf: []u8) FsError!usize {
        const start: usize = @intCast(self.offset);
        if (start >= self.size) return 0;
        const end = @min(start + buf.len, self.size);
        const n = end - start;
        @memcpy(buf[0..n], self.fs.data[self.data_offset + start .. self.data_offset + end]);
        self.offset += n;
        return n;
    }
    fn write(_: *TarFileHandle, _: []const u8) FsError!usize {
        return FsError.PermissionDenied;
    }
    fn seek(self: *TarFileHandle, pos: SeekFrom) FsError!u64 {
        const size: i64 = @intCast(self.size);
        const new_off: i64 = switch (pos) {
            .start => |n| @intCast(n),
            .current => |n| @as(i64, @intCast(self.offset)) + n,
            .end => |n| size + n,
        };
        if (new_off < 0) return FsError.InvalidPath;
        self.offset = @intCast(new_off);
        return self.offset;
    }
    fn stat(self: *TarFileHandle) FsError!Metadata {
        const ms = self.mtime *| 1000;
        return Metadata.fileWithNlink(self.size, self.nlink).withMode(@intCast(self.mode & 0o7777)).withTimes(ms, ms, ms);
    }

    const handle_vtable = FileHandle.VTable{
        .read = hRead,
        .write = hWrite,
        .seek = hSeek,
        .stat = hStat,
        .truncate = vfs.handleTruncateUnsupported,
        .close = hClose,
    };
    fn hRead(p: *anyopaque, buf: []u8) FsError!usize {
        return h_(p).read(buf);
    }
    fn hWrite(p: *anyopaque, buf: []const u8) FsError!usize {
        return h_(p).write(buf);
    }
    fn hSeek(p: *anyopaque, pos: SeekFrom) FsError!u64 {
        return h_(p).seek(pos);
    }
    fn hStat(p: *anyopaque) FsError!Metadata {
        return h_(p).stat();
    }
    fn hClose(p: *anyopaque) void {
        const self = h_(p);
        self.fs.gpa.destroy(self);
    }
    fn h_(p: *anyopaque) *TarFileHandle {
        return @ptrCast(@alignCast(p));
    }
};

fn normPath(path: []const u8, buf: []u8) ?[]const u8 {
    _ = buf;
    if (path.len == 0 or std.mem.eql(u8, path, ".")) return "/";
    return path;
}

/// Normalize a tar entry name to an absolute VFS path (arena/gpa-owned).
fn tarPath(alloc: std.mem.Allocator, name: []const u8) []const u8 {
    if (std.mem.startsWith(u8, name, "./")) {
        const stripped = std.mem.trimEnd(u8, name[1..], "/");
        return alloc.dupe(u8, if (stripped.len == 0) "/" else stripped) catch @panic("OOM");
    } else if (std.mem.startsWith(u8, name, "/")) {
        return alloc.dupe(u8, std.mem.trimEnd(u8, name, "/")) catch @panic("OOM");
    }
    const stripped = std.mem.trimEnd(u8, name, "/");
    return std.fmt.allocPrint(alloc, "/{s}", .{stripped}) catch @panic("OOM");
}
